interfaces/cython/cantera/examples/reactors/combustor.py

```diff
@@ -1,93 +1,83 @@
-equiv_ratio = 0.5
-
-# compute fuel and air mass flow rates
-factor = 0.1
-air_mdot = factor * 9.52 * air_mw
-fuel_mdot = factor * equiv_ratio * fuel_mw
-
-# create and install the mass flow controllers. Controllers m1 and m2 provide
-# constant mass flow rates, and m3 provides a short Gaussian pulse only to
-# ignite the mixture
-m1 = ct.MassFlowController(fuel_in, combustor, mdot=fuel_mdot)
+# Use a variable mass flow rate to keep the residence time in the reactor
+# constant (residence_time = mass / mass_flow_rate). The mass flow rate function
+# can access variables defined in the calling scope, including state variables
+# of the Reactor object (combustor) itself.
 
-# note that this connects two reactors with different reaction mechanisms and
-# different numbers of species. Downstream and upstream species are matched by
-# name.
-m2 = ct.MassFlowController(air_in, combustor, mdot=air_mdot)
+def mdot(t):
+    return combustor.mass / residence_time
 
-# The igniter will use a Gaussian time-dependent mass flow rate.
-fwhm = 0.2
-amplitude = 0.1
-t0 = 1.0
-igniter_mdot = lambda t: amplitude * math.exp(-(t-t0)**2 * 4 * math.log(2) / fwhm**2)
-m3 = ct.MassFlowController(igniter, combustor, mdot=igniter_mdot)
+inlet_mfc = ct.MassFlowController(inlet, combustor, mdot=mdot)
 
-# put a valve on the exhaust line to regulate the pressure
-v = ct.Valve(combustor, exhaust, K=1.0)
+# A PressureController has a baseline mass flow rate matching the 'master'
+# MassFlowController, with an additional pressure-dependent term. By explicitly
+# including the upstream mass flow rate, the pressure is kept constant without
+# needing to use a large value for 'K', which can introduce undesired stiffness.
+outlet_mfc = ct.PressureController(combustor, exhaust, master=inlet_mfc, K=0.01)
 
 # the simulation only contains one reactor
 sim = ct.ReactorNet([combustor])
 
-# take single steps to 6 s, writing the results to a CSV file for later
-# plotting.
-tfinal = 6.0
-tnow = 0.0
-Tprev = combustor.T
-tprev = tnow
-states = ct.SolutionArray(gas, extra=['t','tres'])
-
-while tnow < tfinal:
-    tnow = sim.step()
-    tres = combustor.mass/v.mdot(tnow)
-    Tnow = combustor.T
-    if abs(Tnow - Tprev) > 1.0 or tnow-tprev > 2e-2:
-        tprev = tnow
-        Tprev = Tnow
-        states.append(gas.state, t=tnow, tres=tres)
-
-states.write_csv('combustor.csv', cols=('t','T','tres','X'))
+# Run a loop over decreasing residence times, until the reactor is extinguished,
+# saving the state after each iteration.
+states = ct.SolutionArray(gas, extra=['tres'])
+
+residence_time = 0.1 # starting residence time
+while combustor.T > 500:
+    sim.set_initial_time(0.0) # reset the integrator
+    sim.advance_to_steady_state()
+    print('tres = {:.2e}; T = {:.1f}'.format(residence_time, combustor.T))
+    states.append(combustor.thermo.state, tres=residence_time)
+    residence_time *= 0.9 # decrease the residence time for the next iteration
+
+if __name__ == '__main__':
```

Comment:
I think I confused myself while writing this. The idea I had was "what if you don't have Matplotlib installed?", but after looking at our other examples, there are quite a few where we just assume it's available. And of course, this doesn't actually help in that case. We do have a couple where we put all of the plotting in a `try / except ImportError` block, so that might be a better option here, or we can just go straight for the plotting.

Reply:
I say just go for it. People should have matplotlib installed anyways, otherwise, how do they visualize their results? Export to Excel? 🤢